Adopt safer CPP in AudioWorkletThread, SQLTransactionCoordinator, CSSPendingSubstitutionValue, CDMClearKey, SVGAltGlyphElement

jelee53 opened the pull request:
#### 28f408c82d9e05b448ba4efc98077f12ed144f6b
<pre>
Adopt safer CPP in AudioWorkletThread, SQLTransactionCoordinator, CSSPendingSubstitutionValue, CDMClearKey, SVGAltGlyphElement
<a href="https://bugs.webkit.org/show_bug.cgi?id=302993">https://bugs.webkit.org/show_bug.cgi?id=302993</a>
<a href="https://rdar.apple.com/165248345">rdar://165248345</a>

Reviewed by Per Arne Vollan.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.</a>

No new tests needed.

* Source/WebCore/Modules/webaudio/AudioWorkletThread.cpp:
(WebCore::AudioWorkletThread::createThread):
* Source/WebCore/Modules/webdatabase/SQLTransactionCoordinator.cpp:
(WebCore::SQLTransactionCoordinator::shutdown):
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/css/CSSPendingSubstitutionValue.cpp:
(WebCore::CSSPendingSubstitutionValue::resolveValue const):
* Source/WebCore/platform/encryptedmedia/clearkey/CDMClearKey.cpp:
(WebCore::CDMInstanceSessionClearKey::requestLicense):
* Source/WebCore/platform/graphics/ColorUtilities.h:
(WebCore::colorByModifingEachNonAlphaComponent):

Canonical link: <a href="https://commits.webkit.org/303999@main">https://commits.webkit.org/303999@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3bc131521e5af2a5a320d4ad1846be11fd05f877

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134177 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6688 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45388 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141757 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86239 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/830c25bf-fcdd-4c80-9753-0c30a66e9ea1) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136047 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7226 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6552 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102602 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/69894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/945932f1-6fc6-4808-ba37-ed6c4f1dc87f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137124 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5038 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120265 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83396 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4918 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2557 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/1572 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114087 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38401 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144403 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6358 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38979 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110976 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6440 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5345 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111223 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4781 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116538 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60129 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20734 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6410 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34755 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6256 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69876 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6500 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6364 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/webaudio/AudioWorkletThread.cpp:
```diff
@@ -69,7 +69,7 @@ WorkerDebuggerProxy* AudioWorkletThread::workerDebuggerProxy() const
 Ref<Thread> AudioWorkletThread::createThread()
 {
-    return Thread::create("WebCore: AudioWorklet"_s, [this] {
+    return Thread::create("WebCore: AudioWorklet"_s, [this, protectedThis = Ref { *this }] {
```
This looks like a ref-cycle via `m_thread` to me no? I suspect we're introducing a leak here.

cdumez commented on Source/WebCore/Modules/webdatabase/SQLTransactionCoordinator.cpp:
```diff
@@ -123,8 +123,8 @@ void SQLTransactionCoordinator::shutdown()
         // transaction is interrupted?" at the top of SQLTransactionBackend.cpp.
-        if (info.activeWriteTransaction)
-            info.activeWriteTransaction->notifyDatabaseThreadIsShuttingDown();
+        if (RefPtr activeWriteTransaction = info.activeWriteTransaction)
```
We can even name the variable `transaction` for such a small scope.

cdumez commented on Source/WebCore/css/CSSPendingSubstitutionValue.cpp:
```diff
@@ -33,7 +33,7 @@ namespace WebCore {
 RefPtr<CSSValue> CSSPendingSubstitutionValue::resolveValue(Style::Builder& builder, CSSPropertyID propertyID) const
 {
-    auto cacheValue = [&](auto data) {
+    auto cacheValue = [this, protectedThis = Ref { *this }](auto data) {
```
The lambda parameter to `resolveAndCacheValue()` is marked as `NOESCAPE` so this shouldn't be needed. If static analysis is complaining, I suggest we get rid of this local variable and pass in the lambda directly to `resolveAndCacheValue()`. Then I'm pretty sure the static analysis would be happy with capturing just `this`.

cdumez commented on Source/WebCore/platform/encryptedmedia/clearkey/CDMClearKey.cpp:
```diff
@@ -461,10 +461,7 @@ void CDMInstanceSessionClearKey::requestLicense(LicenseType, KeyGroupingStrategy
         initData = extractKeyIdFromWebMInitData(initData.get());
 
     callOnMainThread(
```
I am very concerned that we may not be on the main thread given the call to `callOnMainThread()` here. In this case, the old code would be safe but your new code would not because you would now be ref'ing this object off the main thread, even though its ref-count is not thread safe (it subclasses RefCounted).

My suggestion is to simplify the code and not capture `this / weakThis` at all and instead capture `sessionID = m_sessionID`. Also note that the original was not completely correct either since it was failing to call `callback` when `weakThis` was null, even though `callback` is a CompletionHandler (which had to get called once before destruction).

cdumez commented on Source/WebCore/platform/graphics/filters/FilterOperation.cpp:
```diff
@@ -184,20 +184,20 @@ bool BasicComponentTransferFilterOperation::transformColor(SRGBA<float>& color)
         return true;
     case Type::Invert: {
         float oneMinusAmount = 1.0f - m_amount;
-        color = colorByModifingEachNonAlphaComponent(color, [&](float component) {
```
We should mark the lambda parameter to `colorByModifingEachNonAlphaComponent` as `NOESCAPE` instead since it runs synchronously.

cdumez commented on Source/WebCore/platform/graphics/filters/FilterOperation.cpp:
```diff
@@ -184,20 +184,20 @@ bool BasicComponentTransferFilterOperation::transformColor(SRGBA<float>& color)
     case Type::Contrast: {
         float intercept = -(0.5f * m_amount) + 0.5f;
-        color = colorByModifingEachNonAlphaComponent(color, [&](float component) {
+        color = colorByModifingEachNonAlphaComponent(color, [this, protectedThis = Ref { *this }, intercept](float component) {
```
ditto.

cdumez commented on Source/WebCore/platform/graphics/filters/FilterOperation.cpp:
```diff
@@ -184,20 +184,20 @@ bool BasicComponentTransferFilterOperation::transformColor(SRGBA<float>& color)
     }
     case Type::Brightness:
-        color = colorByModifingEachNonAlphaComponent(color, [&](float component) {
+        color = colorByModifingEachNonAlphaComponent(color, [this, protectedThis = Ref { *this }](float component) {
```
ditto.

cdumez commented on Source/WebCore/svg/SVGAltGlyphElement.cpp:
```diff
@@ -82,7 +82,7 @@ RenderPtr<RenderElement> SVGAltGlyphElement::createElementRenderer(RenderStyle&&
 {
     // No need to support altGlyph referencing another node inside a shadow tree.
-    auto target = targetElementFromIRIString(getAttribute(SVGNames::hrefAttr, XLinkNames::hrefAttr), document());
+    auto target = targetElementFromIRIString(getAttribute(SVGNames::hrefAttr, XLinkNames::hrefAttr), Ref { document() }.get());
```
`protectedDocument()` ?

pvollan commented on Source/WebCore/Modules/webaudio/AudioWorkletThread.cpp:
```diff
@@ -69,7 +69,11 @@ WorkerDebuggerProxy* AudioWorkletThread::workerDebuggerProxy() const
 Ref<Thread> AudioWorkletThread::createThread()
 {
-    return Thread::create("WebCore: AudioWorklet"_s, [this] {
+    return Thread::create("WebCore: AudioWorklet"_s, [this, weakThis = ThreadSafeWeakPtr { *this }] {
```
Can we avoid capturing `this` now?

cdumez replied:
Right, we should no longer capture `this`

jelee53 committed:
Adopt safer CPP in AudioWorkletThread, SQLTransactionCoordinator, CSSPendingSubstitutionValue, CDMClearKey, SVGAltGlyphElement
https://bugs.webkit.org/show_bug.cgi?id=302993
rdar://165248345

Reviewed by Per Arne Vollan.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.

No new tests needed.

* Source/WebCore/Modules/webaudio/AudioWorkletThread.cpp:
(WebCore::AudioWorkletThread::createThread):
* Source/WebCore/Modules/webdatabase/SQLTransactionCoordinator.cpp:
(WebCore::SQLTransactionCoordinator::shutdown):
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/css/CSSPendingSubstitutionValue.cpp:
(WebCore::CSSPendingSubstitutionValue::resolveValue const):
* Source/WebCore/platform/encryptedmedia/clearkey/CDMClearKey.cpp:
(WebCore::CDMInstanceSessionClearKey::requestLicense):
* Source/WebCore/platform/graphics/ColorUtilities.h:
(WebCore::colorByModifingEachNonAlphaComponent):

Canonical link: https://commits.webkit.org/303999@main

cdumez commented on Source/WebCore/css/CSSPendingSubstitutionValue.cpp:
```diff
@@ -33,17 +33,14 @@ namespace WebCore {
-    };
-
-    if (!m_shorthandValue->resolveAndCacheValue(builder, cacheValue))
+        m_cachedPropertyValues = parsedProperties; }))
```
Normally, we would put `});` on the next line.
